File: back_velora/src/main/java/com/velora/auth/AuthService.java
package com.velora.auth;

import java.util.UUID;

import com.velora.auth.dto.*;
import com.velora.security.JwtService;
import com.velora.user.*;

import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

@Service
public class AuthService {
    private final UserRepository users;
    private final PasswordEncoder passwords;
    private final JwtService jwt;

    public AuthService(UserRepository users, PasswordEncoder passwords, JwtService jwt) {
        this.users = users;
        this.passwords = passwords;
        this.jwt = jwt;
    }

    @Transactional
    public AuthResponse registerCustomer(RegisterRequest request) {
        String email = request.email().trim().toLowerCase();

        if (users.existsByEmailIgnoreCase(email)) {
            throw new ResponseStatusException(HttpStatus.CONFLICT, "Ya existe una cuenta con este correo.");
        }

        UserEntity user = new UserEntity();
        user.setFirstName(request.firstName());
        user.setLastName(request.lastName());
        user.setEmail(email);
        user.setPasswordHash(passwords.encode(request.password()));

        // Regla de negocio: el registro público SIEMPRE crea CUSTOMER.
        user.setRole(UserRole.CUSTOMER);
        user.setCustomerType(CustomerType.B2C);
        user.setStatus(UserStatus.ACTIVE);
        user.setStore(null);

        return jwt.issue(users.save(user));
    }

    @Transactional(readOnly = true)
    public AuthResponse login(LoginRequest request) {
        UserEntity user = users.findByEmailIgnoreCase(request.email().trim())
                .orElseThrow(() -> new ResponseStatusException(
                        HttpStatus.UNAUTHORIZED,
                        "Correo o contraseña incorrectos."
                ));

        if (user.getStatus() != UserStatus.ACTIVE) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN, "La cuenta no está activa.");
        }

        if (!passwords.matches(request.password(), user.getPasswordHash())) {
            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Correo o contraseña incorrectos.");
        }

        return jwt.issue(user);
    }

    @Transactional(readOnly = true)
    public UserProfileResponse profile(UUID id) {
        return users.findById(id)
                .map(UserProfileResponse::from)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Usuario no encontrado."));
    }
}

File: apps/api/src/modules/auth/auth.service.ts
import type {
  ServerRuntimeConfig,
} from "@velora/config";

import type {
  AuthResponse,
  LoginRequest,
  RegisterRequest,
  UserProfile,
} from "@velora/contracts";

import {
  createCustomerUser,
  findUserByEmail,
  findUserById,
  isDatabaseUniqueViolation,
  toUserProfile,
  userEmailExists,
} from "../users/user.repository";

import {
  ApiHttpError,
} from "./http";

import {
  hashPassword,
  issueAuthToken,
  passwordMatches,
} from "./security";

export async function registerCustomer(
  request: RegisterRequest,
  config: ServerRuntimeConfig,
): Promise<AuthResponse> {
  const email =
    request.email
      .trim()
      .toLowerCase();

  if (
    await userEmailExists(
      email,
    )
  ) {
    throw new ApiHttpError(
      409,
      "Ya existe una cuenta con este correo.",
    );
  }

  const passwordHash =
    await hashPassword(
      request.password,
    );

  try {
    const user =
      await createCustomerUser({
        firstName:
          request.firstName,

        lastName:
          request.lastName,

        email,

        passwordHash,
      });

    return issueAuthToken(
      toUserProfile(
        user,
      ),
      config,
    );
  }
  catch (error) {
    if (
      isDatabaseUniqueViolation(
        error,
      )
    ) {
      throw new ApiHttpError(
        409,
        "Ya existe una cuenta con este correo.",
      );
    }

    throw error;
  }
}

export async function login(
  request: LoginRequest,
  config: ServerRuntimeConfig,
): Promise<AuthResponse> {
  const user =
    await findUserByEmail(
      request.email.trim(),
    );

  if (!user) {
    throw new ApiHttpError(
      401,
      "Correo o contraseña incorrectos.",
    );
  }

  if (
    user.status !==
    "ACTIVE"
  ) {
    throw new ApiHttpError(
      403,
      "La cuenta no está activa.",
    );
  }

  const matches =
    await passwordMatches(
      request.password,
      user.passwordHash,
    );

  if (!matches) {
    throw new ApiHttpError(
      401,
      "Correo o contraseña incorrectos.",
    );
  }

  return issueAuthToken(
    toUserProfile(
      user,
    ),
    config,
  );
}

export async function profile(
  id: string,
): Promise<UserProfile> {
  const user =
    await findUserById(
      id,
    );

  if (!user) {
    throw new ApiHttpError(
      404,
      "Usuario no encontrado.",
    );
  }

  return toUserProfile(
    user,
  );
}